### PR TITLE
Profile in rust toolchain

### LINKED ISSUE #2579: rust-toolchain should allow setting the equivalent of `--profile minimal`
**Describe the problem you are trying to solve**

For things like CI that can make do with the minimal profile but also need additional components, I usually do something like this to install the desired toolchain:

```sh
rust_toolchain="$(cat rust-toolchain)"
rustup toolchain install "$rust_toolchain" --profile minimal --component clippy,rustfmt
rustup default "$rust_toolchain"
```

The new TOML format allows this to be skipped since the first invocation of `rustc` / `cargo` will auto-install the toolchain with the required components, *except* that it does so using the default profile and thus also installs `rust-docs`.

**Describe the solution you'd like**

It would be nice if, when rustup installs a new toolchain as a result of a discovered TOML rust-toolchain file, it either defaults to using the minimal profile if `components` is specified (too late now perhaps due to backward compatibility), or the TOML has a `profile` field I can set to `"minimal"`, or some other way.

### PR DESCRIPTION
**Pull request by kinnison**
If I'm right this should fix #2579

@jyn514 Is this what you were expecting / desiring?  If so, you can look at the built artifacts and fetch a test rustup to try.

### CODE REVIEW

File: src/toolchain.rs [hunk 1]
```diff
@@ -685,12 +686,15 @@ impl<'a> DistributableToolchain<'a> {
...
+        profile: Option<Profile>,
     ) -> Result<UpdateStatus> {
         let update_hash = self.update_hash()?;
         let old_date = self.get_manifest().ok().and_then(|m| m.map(|m| m.date));
         InstallMethod::Dist {
             desc: &self.desc()?,
-            profile: self.0.cfg.get_profile()?,
+            profile: profile
+                .map(|p| Ok(p))
+                .unwrap_or_else(|| self.0.cfg.get_profile())?,
```
Comment (jyn514) [inline src/toolchain.rs]:
```suggestion
            profile: profile.map_or_else(Ok, || self.0.cfg.get_profile())?,
```
or if you don't like `map_or_else` :P just changing the closure to `Ok` directly.

File: tests/cli-rustup.rs [hunk 2]
```diff
@@ -1548,6 +1548,35 @@ components = [ "rust-bongo" ]
...
+[toolchain]
+profile = "minimal"
+channel = "nightly"
+"#,
+        )
+        .unwrap();
+        expect_not_stdout_ok(
+            config,
+            &["rustup", "component", "list"],
+            for_host!("rust-docs-{} (installed)"),
```
Comment (jyn514) [inline tests/cli-rustup.rs:1575]: This says 'expect the component not to be present'? Looks good if so :)

**Review by jyn514 — APPROVED** [review]
Just a tiny nit. Thanks so much for adding this!

**Comment by jyn514** [comment]
> If so, you can look at the built artifacts and fetch a test rustup to try.

~~How can I look at the built artifacts?~~ Found it, it was in the GitHub actions screen. I had to be logged in to download the artifact for some reason.

**Comment by jyn514** [comment]
Tested it out and it works great, just like I'd expect :)

File: src/toolchain.rs (hunk 1 above)
Comment (kinnison) [reply src/toolchain.rs]: I keep forgetting that variant constructors are functions, I also was unaware of `.map_or_else()` which I now need to add to my brain.

File: tests/cli-rustup.rs (hunk 2 above)
Comment (kinnison) [reply tests/cli-rustup.rs:1575]: Correct.  it's a tad unintuitive, but the idea is to demonstrate that the profile overrode the user setting.

File: src/config.rs [hunk 3]
```diff
@@ -89,6 +91,10 @@ impl<'a> OverrideCfg<'a> {
             },
             components: file.toolchain.components.unwrap_or_default(),
             targets: file.toolchain.targets.unwrap_or_default(),
+            profile: match file.toolchain.profile {
+                Some(name) => Some(dist::Profile::from_str(&name)?),
+                None => None,
+            },
```
Comment (toothbrush7777777) [inline src/config.rs]: You could use `Option::map` here, right?

Comment (kinnison) [reply src/config.rs]: Mmm, a combination of `.map()` and `.transpose()` might make that work.

**Comment by toothbrush7777777** [comment]
Yeah, I was thinking of `and_then`, not `map`.

File: src/config.rs (hunk 3 above)
Comment (toothbrush7777777) [reply src/config.rs]: Yeah, I was thinking of `and_then`, not `map`.

**Comment by kinnison** [comment]
I've worked through the code suggestions and if this goes green I'll merge.

**Pull request merged**